mysql/mysqldump.go

```diff
@@ -0,0 +1,578 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	driver "github.com/pingcap/tidb/types/parser_driver"
+)
+
+var valuesRegexp = regexp.MustCompile("\\((.*?)\\)")
+var insertRegexp = regexp.MustCompile("INSERT\\sINTO\\s(.*?)\\sVALUES\\s")
+
+// MysqlSpatialDataTypes is an array of all mysql spatial data types.
+var MysqlSpatialDataTypes = []string{"geometrycollection", "multipoint", "multilinestring", "multipolygon", "point", "linestring", "polygon", "geometry"}
+
+// ProcessMySQLDump reads mysqldump data from r and does schema or data conversion,
+// depending on whether conv is configured for schema mode or data mode.
+// In schema mode, ProcessMySQLDump incrementally builds a schema (updating conv).
+// In data mode, ProcessMySQLDump uses this schema to convert MySQL data
+// and writes it to Spanner, using the data sink specified in conv.
+func ProcessMySQLDump(conv *internal.Conv, r *internal.Reader) error {
+	for {
+		startLine := r.LineNumber
+		startOffset := r.Offset
+		b, stmts, err := readAndParseChunk(conv, r)
+		if err != nil {
+			return err
+		}
+		for _, stmt := range stmts {
+			isInsert := processStatement(conv, stmt)
+			internal.VerbosePrintf("Parsed SQL command at line=%d/fpos=%d: %d stmts (%d lines, %d bytes) Insert Statement=%v\n", startLine, startOffset, 1, r.LineNumber-startLine, len(b), isInsert)
+			if isInsert {
+				switch ddlstmt := stmt.(type) {
+				case *ast.InsertStmt:
+					processInsertStmt(conv, ddlstmt)
+				}
+			}
+		}
+		if r.EOF {
+			break
+		}
+	}
+	if conv.SchemaMode() {
+		schemaToDDL(conv)
+		conv.AddPrimaryKeys()
+	}
+	return nil
+}
+
+// readAndParseChunk parses a chunk of mysqldump data, returning the bytes read,
+// the parsed AST (nil if nothing read), error and whether we've hit end-of-file.
+func readAndParseChunk(conv *internal.Conv, r *internal.Reader) ([]byte, []ast.StmtNode, error) {
+	var l [][]byte
+	for {
+		b := r.ReadLine()
+		l = append(l, b)
+		// If we see a semicolon or eof, we're likely to have a command, so try to parse it.
+		// Note: we could just parse every iteration, but that would mean more attempts at parsing.
+		if strings.Contains(string(b), ";") || r.EOF {
+			n := 0
+			for i := range l {
+				n += len(l[i])
+			}
+			s := make([]byte, n)
+			n = 0
+			for i := range l {
+				n += copy(s[n:], l[i])
+			}
+			query := string(s)
+			tree, _, err := parser.New().Parse(query, "", "")
+			if err == nil {
+				return s, tree, nil
+			}
+			newTree, err1 := handleParseError(conv, query, err, l)
+			if err1 == false {
```

Comment:
No need to compare to false.

if !err1 { ...}

is simpler and clearer.

Also, I'm not sure about using err1 for a bool value -- go uses variables 'err' for errors very systematically.

For boolean return values that represent status, its more common to use ok, as is done with golang maps (https://blog.golang.org/maps).

So maybe:

newTree, ok := handleParseError(...)
if ok {
}

Of course, that means you'll need to change the return values in handleParseError (true becomes false etc).

Reply:
Done.
https://github.com/cloudspannerecosystem/harbourbridge/blob/99db369b2b826fc24ffbb760f3b3700cd0e04b0f/mysql/mysqldump.go#L92-L95